### ISSUE TITLE
v8 error: JavaScript heap out of memory when generating types for generic types with inline fragments

### ISSUE THREAD

**Issue by SeppahBaws** [HoudiniGraphql/houdini#886]
### Describe the bug

In our project at work today, I wrote a query that looks somewhat similar to this:
```gql
query MyAnimalQuery {
  animal(id: "some_cat") {
    id

    ...AnimalDetails

    ... on Cat {
      ...CatThings
    }
  }
}

fragment AnimalDetails on Animal {
  id
  name
  description

  # Need this so that Houdini can generate the __typename properly
  ... on Cat {
    id
  }
}

fragment CatThings on Cat {
  id
  name
  nrOfGlassesKnockedOver

  ...AnimalDetails
}
```
Each fragment is in a separate component, where they're used to display the proper data in the UI.

Upon running `houdini generate`, it hangs for a while, eats up 4GB of RAM, and then v8 crashes because it ran out of heap memory:
```
<--- Last few GCs --->

[7752:0000021720CE3E20]    33580 ms: Mark-sweep 4054.5 (4128.4) -> 4044.1 (4134.1) MB, 2030.9 / 0.0 ms  (average mu = 0.195, current mu = 0.073) allocation failure scavenge might not succeed
[7752:0000021720CE3E20]    35686 ms: Mark-sweep 4059.9 (4134.1) -> 4047.5 (4137.4) MB, 2094.9 / 0.0 ms  (average mu = 0.107, current mu = 0.005) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF73BEF0AAF v8::internal::CodeObjectRegistry::~CodeObjectRegistry+124015
 2: 00007FF73BE7C866 v8::internal::wasm::WasmCode::safepoint_table_offset+64182
 3: 00007FF73BE7D8E2 v8::internal::wasm::WasmCode::safepoint_table_offset+68402
 4: 00007FF73C7B1CE4 v8::Isolate::ReportExternalAllocationLimitReached+116
 5: 00007FF73C79C2AD v8::SharedArrayBuffer::Externalize+781
 6: 00007FF73C63F88C v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1468
 7: 00007FF73C64C539 v8::internal::Heap::PublishPendingAllocations+1129
 8: 00007FF73C64950A v8::internal::Heap::PageFlagsAreConsistent+2842
 9: 00007FF73C63C169 v8::internal::Heap::CollectGarbage+2137
10: 00007FF73C63A320 v8::internal::Heap::AllocateExternalBackingStore+2000
11: 00007FF73C65EEA6 v8::internal::Factory::NewFillerObject+214
12: 00007FF73C391645 v8::internal::DateCache::Weekday+1797
13: 00007FF73C83F9C1 v8::internal::SetupIsolateDelegate::SetupHeap+494417
14: 0000021722D4A7A9
```

Commenting out the `... on Cat` in the `AnimalDetails` fragment fixes it.
Funnily enough, if I change it to `... on Dog`, houdini generate works just fine.

This might be because of me slightly mis-using the powers of GraphQL, but the query itself works just fine on the GraphQL server, so I thought I'd report it.
In the end I worked around this by fetching the data in a cleaner way, so it's not a blocker atm.

### Severity

serious, but I can work around it

### Steps to Reproduce the Bug

1. Clone the linked repo, on the `bug-repro/js-out-of-memory` branch (example code is in the e2e/sveltekit/src/routes/js-out-of-memory directory)
2. Run `npx houdini generate` in the e2e/sveltekit project
3. Wait and see the error pop up after a while (also note the RAM usage in task manager)
4. Comment out the `... on Cat` in the `AnimalDetails` fragment, and see houdini generate completing successfully

### Reproduction

https://github.com/SeppahBaws/houdini/tree/bug-repro/js-out-of-memory/e2e/sveltekit/src/routes/js-out-of-memory

**Comment by AlecAivazis** [comment]
Oh no! Well, thanks for reporting it (and for the reproduction)!

I'm glad it's not a blocker but I'll find some time to look at this as soon as I can

**Comment by SeppahBaws** [comment]
> Just to confirm - this is with a `0.20.x` release, right?

I tried with both the `0.20.x` and the latest `next` branch, and it happens in both. The repro is based off the `next` branch

**Comment by AlecAivazis** [comment]
I have a fix 🥳 Mind opening a PR from that branch so I can push my changes onto your fork?

**Comment by AlecAivazis** [comment]
Nvm, I was able to open up #889 and save your contribution credit 👍

**Comment by AlecAivazis** [comment]
This should be fixed in `v1.0.0-next.13`

**Comment by SeppahBaws** [comment]
Just confirmed, it seems to be fixed in my fork with the `v1.0.0-next.13` version! 🎉
